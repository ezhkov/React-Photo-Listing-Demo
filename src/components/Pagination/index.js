import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { getPageNumbers, generateNavItems } from "../../utils";
import {
  Pagination,
  PageItem,
  PageLink,
} from 'mdbreact';

class PaginationSelf extends Component {
  constructor(...args) {
    super(...args);
    this.state = {
      page: 1,
      itemNumbers: 10,
      navigationItems: [1],
    };
  }

  componentDidMount = () => {
    this.pageCalc();
  };

  componentDidUpdate = (prevProps) => {
    if (prevProps.totalCards !== this.props.totalCards) this.pageCalc();
  }

  pageCalc = () => {
    const { perPage, totalCards } = this.props;
    const { page, itemNumbers} = this.state;

    const navigationItems = generateNavItems(page, perPage, totalCards, itemNumbers);

    this.setState({
      navigationItems,
    });
  }

  handleNavigationPrevClick = () => {
    const { onNavigationPrevClick } = this.props;
    const { page } = this.state;
    if (page - 1 < 1) return false;
    else {
      this.setState({
        page: page - 1,
      }, onNavigationPrevClick);
    }
  }

  handleNavigationNextClick = () => {
    const { totalCards, perPage, onNavigationNextClick } = this.props;
    const { page } = this.state;
    const i = getPageNumbers(page, perPage, totalCards);
    if (page + 1 >= i) return false;
    else {
      
      this.setState({
        page: page + 1,
      }, onNavigationNextClick);
    }
  }

  handleNavigationClick = (item) => {
    const { onNavigationClick } = this.props;
    const { page } = this.state;
    if (page === item) return false;
    else {
      this.setState({
        page: item,
      }, onNavigationClick(item));
    }
  }

  render() {

    const { navigationItems, page, itemNumbers } = this.state;

    return (
      <Pagination className="pagination-lg pagination">
        <PageItem className="pagination__item pagination__item_left">
          <PageLink
            onClick={this.handleNavigationPrevClick}
            className="page-link page-link_left"
            aria-label="Previous"
          >
            <span aria-hidden="true">&laquo;</span>
            <span className="sr-only">Previous</span>
          </PageLink>
        </PageItem>
        {
          navigationItems.map((item, i) => {
            if (i < itemNumbers) {
              return (
                <PageItem className={`pagination__item ${item === page ? 'pagination__item_active' : ''}`} active={item === page} key={item}>
                  <PageLink key={item} onClick={() => this.handleNavigationClick(item)} className={`page-link pagination__link ${item === page ? 'pagination__link_active' : ''}`}>
                    {item}
                  </PageLink>
                </PageItem>);
            }
            return false;
          })
        }
        <PageItem className="pagination__item pagination__item_right">
          <PageLink
            onClick={this.handleNavigationNextClick}
            className="page-link page-link_right"
            aria-label="Next"
          >
            &raquo;
          </PageLink>
        </PageItem>
      </Pagination>
    );
  }
}
PaginationSelf.propTypes = {
  page: PropTypes.number,
  perPage: PropTypes.number,
  totalCards: PropTypes.number,
  onNavigationClick: PropTypes.func,
  onNavigationPrevClick: PropTypes.func,
  onNavigationNextClick: PropTypes.func,
};
PaginationSelf.defaultProps = {
  page: 1,
  perPage: 20,
  totalCards: 0,
  onNavigationClick: () => { },
  onNavigationPrevClick: () => { },
  onNavigationNextClick: () => { },
};
export default PaginationSelf;
