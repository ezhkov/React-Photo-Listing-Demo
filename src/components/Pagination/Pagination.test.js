import React from 'react';
import Enzyme, { shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Pagination from '.';

Enzyme.configure({ adapter: new Adapter() });

describe('Test of component of Pagination', () => {
  it('Test state PhotoCard component', () => {
    const pagination = mount(<Pagination />)

  });

  it('Test snapshot Pagination component', () => {
    const pagination = shallow(<Pagination />);
    expect(pagination).toMatchSnapshot();
  });

  it('Test snapshot Pagination component', () => {
    const props = {
      page: 3,
      perPage: 20,
      totalCards: 100,
    }

    const pagination = mount(<Pagination {...props} />);
    //console.log('1111', pagination.find('a.pagination__link_active').text());
    //expect(pagination.find('a.pagination__link').length).toEqual(7);
    //expect(pagination.find('a.pagination__link_active').text()).toEqual('3');

    //pagination.find('a.pagination__link').at(1).simulate('click');
    //expect(pagination.find('a.pagination__link_active').text()).toEqual('1');
     //pagination.instance().handleNavigationClick(1)
    //console.log(pagination.debug());
  });
});