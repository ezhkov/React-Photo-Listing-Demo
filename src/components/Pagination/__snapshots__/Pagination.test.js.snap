// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test of component of Pagination Test snapshot Pagination component 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <PaginationSelf
    onNavigationClick={[Function]}
    onNavigationNextClick={[Function]}
    onNavigationPrevClick={[Function]}
    page={1}
    perPage={20}
    totalCards={0}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): false,
  Symbol(enzyme.__nodes__): Array [
    false,
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
